# Resources
The list is original organised by Lukasz Kurgan's lab (Virginia Commonwealth University, College of Engineering, Copmuter Science).


This is a partial and continually evolving list.
* `AAindex`: repository of physicochemical and biochemical properties of amino acids.
* `BindingDB`: repository of binding affinities for protein-drug interactions.
* `CATH`: structural classification of proteins.
* `D2P2`: repository of putative annotations of intrinsic disorder in proteins.
* `DisProt`: repository of native annotations of intrinsic disorder in proteins.
* `DrugBank`: repository of drugs and drug targets.
* `ELM`: repository of functional sites in proteins.
* `InterPro`: classification and prediction of protein domains and functions.
* `KEGG`: Kyoto encyclopedia of genes and genomes.
* `mentha`: repository of curated protein-protein interactions.
* `MobiDB`: repository of intrinsic disorder and mobility in protein.
* `PDB` (Protein Data Bank): repository of secondary and tertiary protein structures.
* `PDID`: repository of protein-drug interaction in structural human proteome.
* `PISCES`: server for culling of protein sequences.
* `PROSITE`: repository of protein domains, families and functional sites.
* `RefSeq`: repository of non-redundant set of reference protein sequences.
* `SCOPe`: structural classification of proteins.
* `UniProt`: repository of protein sequences and their functional annotations.